### PR TITLE
Cordis to elasticsearch [prototype for EURITO]

### PR DESCRIPTION
**Pull request by jaklinger**
Refers to issue #222

### CODE REVIEW

**Comment by jaklinger** [comment]
Copied from issue #222:
- [ ] Identify whether any data processing needs to be redone to existing Cordis data
- [ ] If so, do it
- [ ] Transfer existing Cordis data to ES
- [ ] Do quality checks
- [ ] Recollect latest Cordis data
- [ ] Transfer Cordis data to ES
- [ ] ES2ES (Nesta --> EURITO)
- [ ] Do quality checks

Also: 
- [ ] Hotfix of docs and build failure of dev branch
- [ ] Merge stable branch into Master

**Comment by bishax** [comment]
Where can I find the appropriate `elasticsearch.config` with entries for cordis?

```python
Runtime error:
Traceback (most recent call last):
  File "/home/alex/miniconda3/envs/nesta/lib/python3.6/site-packages/luigi/worker.py", line 199, in run
    new_deps = self._run_get_new_deps()
  File "/home/alex/miniconda3/envs/nesta/lib/python3.6/site-packages/luigi/worker.py", line 139, in _run_get_new_deps
    task_gen = self.task.run()
  File "/home/alex/nesta/nesta/core/luigihacks/autobatch.py", line 114, in run
    job_params = self.prepare()
  File "/home/alex/nesta/nesta/core/luigihacks/sql2estask.py", line 78, in prepare
    aliases=self.aliases)
  File "/home/alex/nesta/nesta/core/orms/orm_utils.py", line 118, in setup_es
    es_config = get_config('elasticsearch.config', key)
  File "/home/alex/nesta/nesta/core/luigihacks/misctools.py", line 24, in get_config
    return dict(config[header])
  File "/home/alex/miniconda3/envs/nesta/lib/python3.6/configparser.py", line 959, in __getitem__
    raise KeyError(key)
KeyError: 'cordis-eu_dev'
```

**Comment by bishax** [comment]
(I've added an entry to https://eu-west-2.console.aws.amazon.com/es/home?region=eu-west-2#domain:resource=eurito-dev;action=dashboard;tab=TAB_OVERVIEW_ID in my local copy)

File: nesta/core/orms/arxiv-eu_es_config.json [hunk 1]
```diff
@@ -53,7 +53,7 @@
                     },
                     "type": "text"
                 },
-                "terms_nuts0_article": {
+                "terms_fieldsOfStudy_article": {
```
Comment (bishax) [inline nesta/core/orms/arxiv-eu_es_config.json:56]: Note: unrelated changes

File: nesta/core/routines/eurito_es/es_root.py [hunk 2]
```diff
@@ -54,12 +55,14 @@ def requires(self):
...
                               intermediate_bucket=S3_BUCKET)
 
-        #params = (('arxiv', 'article', Article.id),
-        #          ('crunchbase', 'company', Organization.id),
-        #          ('patstat', 'patent', ApplnFamily.docdb_family_id))
+        params = (('arxiv', 'article', Article.id),
+                  ('crunchbase', 'company', Organization.id),
+                  ('patstat', 'patent', ApplnFamily.docdb_family_id),
+                  ('cordis', 'project', Project.rcn),)
+        #params = (('arxiv', 'article', Article.id),)
```
Comment (bishax) [inline nesta/core/routines/eurito_es/es_root.py:62]: Cleanup?

File: nesta/packages/cordis/cordis_api.py [hunk 3]
```diff
@@ -72,7 +72,11 @@ def hit_api(api='', rcn=None, content_type=None):
     r = requests.get(url, params={'lang': 'en',
                                   'rcn': rcn,
                                   'paramType': 'rcn',
-                                  'contenttype': content_type})
+                                  'contenttype': content_type},
+                     headers={'User-Agent':('Mozilla/5.0 (Linux; Android 6.0; '
```
Comment (bishax) [inline nesta/packages/cordis/cordis_api.py:76]: Declare user-agent as constant at top of file?

File: nesta/packages/cordis/cordis_api.py [hunk 4]
```diff
@@ -72,7 +72,11 @@ def hit_api(api='', rcn=None, content_type=None):
...
+                                            'Nexus 5 Build/MRA58N) AppleWebKit/537.36 '
+                                            '(KHTML, like Gecko) Chrome/81.0.4044.92 '
+                                            'Mobile Safari/537.36')})
     if (r.status_code == 500 and
```
Comment (bishax) [inline nesta/packages/cordis/cordis_api.py:80]: What case does this capture?

File: nesta/packages/geo_utils/lookup.py [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import requests
 import pandas as pd
+from io import StringIO
```
Comment (bishax) [inline nesta/packages/geo_utils/lookup.py:3]: Note: unrelated changes

File: sort_all_json.py [hunk 6]
```diff
@@ -6,6 +6,10 @@
         if 'task_chain' in str(fname):
             continue
         with open(fname) as f:
-            js = json.load(f)
+            try:
```
Comment (bishax) [inline sort_all_json.py:9]: Note: unrelated changes (Running this also made it hard to verify whether `tier1.json` had been sorted or sorted and had entries added)

File: nesta/core/batchables/eurito/cordis_eu/run.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        from nesta.core.orms.orm_utils import setup_es
+        es, es_config = setup_es('dev', True, True,
+                                 dataset='cordis-eu')
+        environ = {'config': ('/home/ec2-user/nesta-cordis2es/nesta/'
+                              'core/config/mysqldb.config'),
+                   'batch_file' : ('cordis-eu_EURITO-ElasticsearchTask-2020-04-10-True-15865345336407135.json'),
+                   'db_name': 'dev',
+                   'bucket': 'nesta-production-intermediate',
+                   'done': "False",
+                   'outinfo': ('https://search-eurito-dev-'
```
Comment (bishax) [inline nesta/core/batchables/eurito/cordis_eu/run.py:121]: If we can expose this information here, why can't `elasticsearch.config` be a committed file?

File: README.rst [hunk 8]
```diff
@@ -18,11 +17,12 @@ on `our GitHub repo`_, you will find complete documentation at our `Read the Doc
 .. _Read the Docs site: https://nesta.readthedocs.io/
 .. _our GitHub repo: https://github.com/nestauk/nesta
 
-..    :build-status:
-..    :imagehttps://img.shields.io/travis/rtfd/readthedocs.org.svg?style=flat
-..    :alt: build status
-..    :scale: 100%
-..    :targethttps://travis-ci.org/rtfd/readthedocs.org
+..
```
[the remark this answers is not in the document]
Comment (jaklinger) [reply README.rst:20]: ah, I made a dummy change to set up the draft PR. I'll revert back.

File: nesta/core/batchables/cordis/cordis_api/run.py [hunk 9]
```diff
@@ -109,7 +109,7 @@ def run():
 
     set_log_level(True)    
     if 'BATCHPAR_config' not in os.environ:
-        os.environ['BATCHPAR_batch_file'] = ('Cordis-2019-09-05-True-1567679610283401.json')
```
[the remark this answers is not in the document]
Comment (jaklinger) [reply nesta/core/batchables/cordis/cordis_api/run.py:112]: I use the `__name__=="__main__"` for debugging the batch tasks locally, it's quite useful to have an example of the parameters in here for that purpose... although I really hate having configuration in code. Another good reason to switch DAPS models! I think pulling out the date and file number in this case would be questionably beneficial, as the two numbers are tied to each other, and practically speaking I copy & paste this from the AWS Batch GUI anyway.

File: nesta/core/batchables/eurito/cordis_eu/run.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                           strans_kwargs=strans_kwargs,
+                           null_empty_str=True,
+                           coordinates_as_floats=True,
+                           listify_terms=True,
+                           do_sort=False,
+                           ngram_fields=['textBody_description_project'])
+
+    # collect file
+    logging.info('Retrieving project ids')
+    nrows = 20 if test else None
```
[the remark this answers is not in the document]
Comment (jaklinger) [reply nesta/core/batchables/eurito/cordis_eu/run.py:73]: good catch, thanks

File: nesta/core/batchables/eurito/cordis_eu/run.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+            row['year'] = start_year if start_year is not None else end_year
+            
+            
+            _desc = row.pop('project_description')
+            _obj = row.pop('objective')
+            row['description'] = f'Description:\n{_desc}\n\nObjective:\n{_obj}'
+            row['link'] = f'https://cordis.europa.eu/project/id/{row["rcn"]}'
+
+            uid = row.pop('rcn')
+            _row = es.index(index=es_index, doc_type=es_type,
```
[the remark this answers is not in the document]
Comment (jaklinger) [reply nesta/core/batchables/eurito/cordis_eu/run.py:123]: removal, thanks

File: nesta/core/batchables/eurito/cordis_eu/run.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                            id=uid, body=row)
+            if not count % 1000:
+                logging.info(f"{count} rows loaded to "
+                             "elasticsearch")
+
+    logging.warning("Batch job complete.")
```
[the remark this answers is not in the document]
Comment (jaklinger) [reply nesta/core/batchables/eurito/cordis_eu/run.py:129]: legacy testing, thanks - removed

File: nesta/core/batchables/eurito/cordis_eu/run.py (hunk 7 above)
Comment (jaklinger) [reply nesta/core/batchables/eurito/cordis_eu/run.py:144]: it's actually unnecessary for me hard code this - the info is already in `es_config`, so I've adapted accordingly

regarding why `elasticsearch.config` isn't public - it's kind of related to not making any of our db info public (obviously I screwed up in this case...)

File: nesta/core/batchables/eurito/cordis_eu/run.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+
+if __name__ == "__main__":
+    set_log_level()
+    if 'BATCHPAR_outinfo' not in os.environ:
+        from nesta.core.orms.orm_utils import setup_es
+        es, es_config = setup_es('dev', True, True,
+                                 dataset='cordis-eu')
+        environ = {'config': ('/home/ec2-user/nesta-cordis2es/nesta/'
```
[the remark this answers is not in the document]
Comment (jaklinger) [reply nesta/core/batchables/eurito/cordis_eu/run.py:138]: thanks, remove user specific stuff

File: nesta/core/orms/arxiv-eu_es_config.json (hunk 1 above)
Comment (jaklinger) [reply nesta/core/orms/arxiv-eu_es_config.json:56]: I ran the `sort_all_json` script, sorry. at least all the jsons are sorted :)

File: nesta/core/routines/eurito_es/es_root.py (hunk 2 above)
Comment (jaklinger) [reply nesta/core/routines/eurito_es/es_root.py:62]: yes, thanks

File: sort_all_json.py (hunk 6 above)
Comment (jaklinger) [reply sort_all_json.py:9]: sorry, i'll segment my PRs next time

**Comment by jaklinger** [comment]
Thanks @bishax, great feedback. if you could cast an eye over this again then should be good to go! I'll make the next PR are bit more concise.
